[SPARK-32180][PYTHON][DOCS] Installation page of Getting Started in PySpark documentation

rohitmishr1484 opened the pull request:
### What changes were proposed in this pull request?
This PR proposes to add getting started- installation to new PySpark docs.

### Why are the changes needed?
Better documentation.

### Does this PR introduce _any_ user-facing change?
No. Documentation only.

### How was this patch tested?
Generating documents locally.

Commit:
Performed suggested changes to getting started/installation page

AmplabJenkins commented:
Can one of the admins verify this patch?

rohitmishr1484 commented:
Hi @HyukjinKwon,

As per your suggestion, I have made 

the necessary changes. Please have a look at it. 

- This is a new pull request on the same topic (installation page), as you rightly pointed the earlier got messed up. Thanks for pointing that out. I am pretty sure it's a mistake from my end since I tried updating a forked repo for the first time. I am attaching a pdf copy of my earlier pull request since it contains all your earlier review and comments (see last few pages of the pdf)

[PR.pdf](https://github.com/apache/spark/files/5169679/PR.pdf)

Please let me know your thoughts and apologize for making you review this document multiple times.

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #128271 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128271/testReport)** for PR 29640 at commit [`d32ece1`](https://github.com/apache/spark/commit/d32ece122ea89030dd947d56f824f4d09740de7d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32899/
Test PASSed.

SparkQA commented:
**[Test build #128271 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128271/testReport)** for PR 29640 at commit [`d32ece1`](https://github.com/apache/spark/commit/d32ece122ea89030dd947d56f824f4d09740de7d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128271/
Test PASSed.

HyukjinKwon commented on python/docs/source/getting_started/index.rst:
```diff
@@ -20,7 +20,10 @@
 
-    quickstart
+    installation
+	quickstart
```
Let's use 4 spaces instead of a tab

HyukjinKwon commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+============= ========================= ====================
+Package       Minimum supported version Note
+============= ========================= ====================
+`pandas`      0.23.2                    Optional
```
Let's mention that this is optional for SQL component

HyukjinKwon commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+Package       Minimum supported version Note
+============= ========================= ====================
+`pandas`      0.23.2                    Optional
+`NumPy`       1.7                       Optional
```
and this is required to ML component (but optional in PySpark if you don't use ML).

HyukjinKwon commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+`Py4J`        0.10.9                    Required
+============= ========================= ====================
+
+**Note**: A prerequisite for PySpark installation is the availability of ``JAVA 8 or 11`` and ``JAVA_HOME`` properly set.
```
Let's don't codeblock ``` ``JAVA 8 or 11`` ```. I would just write it as Java 8 or 11.

srowen replied:
Just say Java 8 or later

HyukjinKwon commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+
+.. code-block:: bash
+
+	source activate pyspark_env
```
Let's don't use tab in this file.

HyukjinKwon commented:
Looks fine otherwise. @holdenk and @srowen can you take a look when you guys are available?

srowen commented on python/docs/source/getting_started/index.rst:
```diff
@@ -20,7 +20,10 @@
 Getting Started
 ===============
 
+This page lists an overview of the basic steps required to setup & get started with PySpark.
```
"This page summarizes the basic steps ..."
& -> and

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+Installation
+============
+
+The official release channel is to download it from `the Apache Spark website <https://spark.apache.org/downloads.html>`_.
```
"Official releases are available from ..."

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+============
+
+The official release channel is to download it from `the Apache Spark website <https://spark.apache.org/downloads.html>`_.
+Alternatively, you can also install it via pip from PyPI.  PyPI installation is usually to use
```
remove "also"
pip -> `pip`
"is usually for standalone..."

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+
+	conda create -n pyspark_env 
+
+After the virtual environment is created, it should be visible under the list of conda environments which can be seen using the following command:
```
I would use Conda or `conda` consistently, depending on whether you mean the product or command

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+
+	conda activate pyspark_env
+
+In lower Conda version, the following command might be used:
```
In earlier Conda versions ... should be used:
(earlier than what?)

rohitmishr1484 replied:
Hi @srowen,

Based on the conda documentation the answer is "Conda version before 4.4". 
[The Source for this info is ](https://docs.conda.io/projects/conda/en/latest/release-notes.html#id208)

> conda **activate:** The logic and mechanisms underlying environment activation have been reworked. With conda 4.4, conda activate and conda deactivate are now the preferred commands for activating and deactivating environments. You'll find they are much more snappy than the source activate and source deactivate commands from previous conda versions. The conda activate command also has advantages of (1) being universal across all OSes, shells, and platforms, and (2) not having path collisions with scripts from other packages like Python virtualenv's activate script.

srowen replied:
Right, I'm suggesting saying that here.

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+
+	source activate pyspark_env
+
+PySpark installation using ``pip`` under Conda environment is official. 
```
What does this mean?

rohitmishr1484 replied:
The information this sentence conveys is - 

- Using Conda to install PySpark via conda-forge is not the official way to do it.
- But If we create a Conda environment using ``conda`` and then install PySpark in the new environment then although it's an indirect way of installation but meets the official implementation guideline.

I am fine with removing this line if you and @HyukjinKwon think it's implicit and not required to be mentioned explicitly. Thanks.

srowen replied:
What do you mean 'official' here?

rohitmishr1484 replied:
@srowen, 

As per my understanding "official" implies the most updated source used to setup PySpark. Based on this understanding- PyPI is considered official whereas Conda-forge is not. Please feel free to correct me if there is a caveat in my understanding. Thanks.

HyukjinKwon replied:
Let's remove this line

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+
+PySpark installation using ``pip`` under Conda environment is official. 
+
+PySpark can be installed in this newly created environment using PyPI as shown before:
```
Why repeat this?

rohitmishr1484 replied:
This was added to make sure new users understand the entire flow under one section but we can certainly remove this statement and link the step to the "Using PyPI" section.  Is that fine? 

> Proposed change: PySpark can be installed in this newly created environment using the command specified in `Using PyPI <#using-pypil>`_

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+Official Release Channel
+------------------------
+
+Different flavor of PySpark is available in `the official release channel <https://spark.apache.org/downloads.html>`_.
```
flavor -> flavors
What do flavors mean here?

rohitmishr1484 replied:
![image](https://user-images.githubusercontent.com/66092034/92303074-44b07c00-ef8f-11ea-9433-b090353fc1fd.png)

In the above image Step 1 & 2 allows us to choose different combinations and the file to be downloaded in Step 3 changes accordingly. Depending on the user's requirement any combination can be selected and thus by "flavor" I mean to denote this.

Would you suggest something else or based on the above explanation, "flavor" word can be used and it doesn't mislead the reader?

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+
+    tar xzvf spark-3.0.0-bin-hadoop2.7.tgz
+
+An important step is to ensure ``SPARK_HOME`` environment variable points to the directory where the code has been extracted. 
```
Ensure the `SPARK_HOME` ...

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+    tar xzvf spark-3.0.0-bin-hadoop2.7.tgz
+
+An important step is to ensure ``SPARK_HOME`` environment variable points to the directory where the code has been extracted. 
+The next step is to properly define ``PYTHONPATH`` such that it can find the PySpark and 
```
Just: Define `PYTHON_PATH` such that ...

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+Installing from Source
+----------------------
+
+To install PySpark from source, refer `Building Spark <https://spark.apache.org/docs/latest/building-spark.html>`_.
```
refer to ...

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+
+To install PySpark from source, refer `Building Spark <https://spark.apache.org/docs/latest/building-spark.html>`_.
+
+Steps for defining ``PYTHONPATH`` is same as described in `Official Release Channel <#official-release-channel>`_. 
```
Refer to ... for steps to define `PYTHON_PATH`

HyukjinKwon commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,119 @@
+`Py4J`        0.10.9                    Required
+============= ========================= ====================
+
+**Note**: A prerequisite for PySpark installation is the availability of ``JAVA 8 or 11`` and ``JAVA_HOME`` properly set.
```
I would also add this note as well:

```
If you are using JDK 11, you should set ``-Dio.netty.tryReflectionSetAccessible=true`` for Arrow related
features. See also `Downloading <https://spark.apache.org/docs/latest/#downloading>`_.
```

rohitmishr1484 replied:
@HyukjinKwon, I have slightly reworded it while adding it in the note. If you feel it doesn't convey the intended thought, please let me know and I will revert it to the original suggestion. Thanks.

Commit:
1. Modified as per suggestions from reviewer

SparkQA commented:
**[Test build #128318 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128318/testReport)** for PR 29640 at commit [`45ce72e`](https://github.com/apache/spark/commit/45ce72ec9cd330f1e917ae4307035cf213b9c344).

rohitmishr1484 commented:
@HyukjinKwon and @srowen, Thanks for your review.

@HyukjinKwon, I have updated all the changes you suggested.
@srowen, I have added comments corresponding to the 3 questions you had and updated other suggested changes. 

I will update other changes (corresponding to the comments) once both of you give a green signal.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/32944/
Test PASSed.

SparkQA commented:
**[Test build #128318 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128318/testReport)** for PR 29640 at commit [`45ce72e`](https://github.com/apache/spark/commit/45ce72ec9cd330f1e917ae4307035cf213b9c344).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128318/
Test PASSed.

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,120 @@
+Using Conda  
+-----------
+
+Conda is an open-source package management and environment management system which is a part of `Anaconda <https://docs.continuum.io/anaconda/>`_ distribution. It is both cross-platform and language agnostic.
```
part of _the_

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,120 @@
+
+    conda activate pyspark_env
+
+In Conda version earlier than 4.4, the following command might be used:
```
should be used

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,120 @@
+
+    pip install pyspark
+
+`PySpark at Conda <https://anaconda.org/conda-forge/pyspark>`_ is not the official release.
```
Same, I don't think this matters

HyukjinKwon replied:
I would just say, for example:

Note that `PySpark at Conda <https://anaconda.org/conda-forge/pyspark>`_ is available but not necessarily synced with PySpark release cycle because it is maintained by the community separately.

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,120 @@
+
+Ensure the ``SPARK_HOME`` environment variable points to the directory where the code has been extracted. 
+Define ``PYTHONPATH`` such that it can find the PySpark and 
+Py4J under ``$SPARK_HOME/python/lib``, one example of doing this is shown below:
```
(By the way I think you need just single back-ticks?)
Start a new sentence at "One example"

HyukjinKwon replied:
oh fyi double backticks here make it like a code block. single backtick makes italic for some reasons.

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,120 @@
+`Py4J`        0.10.9                    Required
+============= ========================= ==========================================================================
+
+**Note**: A prerequisite for PySpark installation is the availability of JAVA 8 or later and ``JAVA_HOME`` properly set. 
```
JAVA -> Java
Nit: you can avoid a lot of these passive sentences. "PySpark requires Java 8 or later, with `JAVA_HOME` properly set" for example.

srowen commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,120 @@
+============= ========================= ==========================================================================
+
+**Note**: A prerequisite for PySpark installation is the availability of JAVA 8 or later and ``JAVA_HOME`` properly set. 
+For using JDK 11, set ``-Dio.netty.tryReflectionSetAccessible=true`` for Arrow related features and refer to `Downloading <https://spark.apache.org/docs/latest/#downloading>`_
```
If using Java 11, set ...

HyukjinKwon commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,120 @@
+============= ========================= ==========================================================================
+`pandas`      0.23.2                    Optional for SQL component
+`NumPy`       1.7                       Required for ML component(Optional in PySpark if ML component is not used)
+`pyarrow`     0.15.1                    Optional
```
pyarrow is also currently only Optional for SQL.

HyukjinKwon commented on python/docs/source/getting_started/installation.rst:
```diff
@@ -0,0 +1,120 @@
+============= ========================= ==========================================================================
+Package       Minimum supported version Note
+============= ========================= ==========================================================================
+`pandas`      0.23.2                    Optional for SQL component
```
I would just write like:

Optional for SQL
Required for ML
Optional for SQL

Seems too long. Sorry for a bit of forth and back here.

Commit:
2. Updated necessary changes

SparkQA commented:
**[Test build #128469 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128469/testReport)** for PR 29640 at commit [`14a97f8`](https://github.com/apache/spark/commit/14a97f83bac5dcddb168dfa990123f11f07833cd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33093/
Test PASSed.

SparkQA commented:
**[Test build #128469 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/128469/testReport)** for PR 29640 at commit [`14a97f8`](https://github.com/apache/spark/commit/14a97f83bac5dcddb168dfa990123f11f07833cd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/128469/
Test PASSed.

rohitmishr1484 commented:
@HyukjinKwon and @srowen,

I have made the suggested changes. Thanks for your review.

srowen commented:
Merged to master

sunchao commented:
@srowen @rohitmishr1484 seems this commit is failing Python linter:

```
Warning, treated as error:
/home/runner/work/spark/spark/python/docs/source/getting_started/installation.rst:20:Duplicate explicit target name: "official release channel".
Makefile:20: recipe for target 'html' failed
make: *** [html] Error 2
```

srowen commented:
Weird - it passed the test run above. Let me take a look and patch it.
